cli/cmd/lql.go

```diff
@@ -19,71 +19,217 @@
 package cmd
 
 import (
+	"fmt"
+	"io"
 	"io/ioutil"
+	"net/http"
+	"strings"
 
 	"github.com/AlecAivazis/survey/v2"
 	"github.com/pkg/errors"
 	"github.com/spf13/cobra"
+
+	"github.com/lacework/go-sdk/api"
 )
 
 var (
-	lqlFile string
+	lqlEnd      string
+	lqlEnv      bool
+	lqlFile     string
+	lqlRepo     bool
+	lqlStart    string
+	lqlURL      string
+	lqlValidate bool
 
-	// lqlCmd represents the lql command
+	// lqlCmd represents the lql parent command
 	lqlCmd = &cobra.Command{
-		Use:    "lql <query>",
-		Hidden: true,
-		Short:  "run an LQL query",
+		Aliases: []string{"lql"},
+		Use:     "query",
+		Short:   "Run and manage LQL queries",
+		Long: `Run and manage LQL queries.
+
+NOTE: This feature is not yet available!`,
+	}
+
+	// lqlRunCmd represents the lql run command
+	lqlRunCmd = &cobra.Command{
+		Use:   "run <query|queryID>",
+		Short: "run an LQL query",
 		Long: `Run an LQL query.
 
-A simple example of an LQL query:
+Run a query via text:
 
-    $ lacework lql 'SimpleLQL_3(RecentComplianceReports Data) {SELECT Data.*}'
+	$ lacework lql run 'SimpleLQL_3(CloudTrailRawEvents e) {SELECT INSERT_ID}' --start <start> --end <end>
 
-NOTE: This feature is not yet available!`,
+Run a query via ID:
+
+	$ lacework lql run MyQuery -e
+
+Start and End times are required to run a query:
+
+1.  Start and End times must be specified in one of the following formats:
+
+	A. ISO 8601 Date and Time
+	B. Epoch time in milliseconds
+
+2. Start and End times must be specified in one of the following ways:
+
+	A.  As StartTimeRange and EndTimeRange in the ParamInfo block within the LQL query
+	B.  As START_TIME_RANGE and END_TIME_RANGE if specifying JSON
+	C.  As --start and --end CLI flags
+	
+3. Start and End time precedence:
+
+	A.  CLI flags take precedence over JSON specifications
+	B.  JSON specifications take precedence over ParamInfo specifications`,
 		Args: cobra.MaximumNArgs(1),
-		RunE: runLQLQuery,
+		RunE: runQuery,
+	}
+
+	// create a slice of the cobra.Command pointers
+	// which need query "text" as input
+	lqlQueryCommands []*cobra.Command = []*cobra.Command{
+		lqlCreateCmd,
+		lqlRunCmd,
+		lqlUpdateCmd,
+		lqlValidateCmd,
 	}
 )
 
 func init() {
 	// add the lql command
 	rootCmd.AddCommand(lqlCmd)
 
-	// file flag to specify a query from disk
-	lqlCmd.Flags().StringVarP(&lqlFile,
-		"file", "f", "",
-		"path to an LQL query to run",
+	// add sub-commands to the lql command
+	lqlCmd.AddCommand(lqlRunCmd)
+
+	// for commands that take query "text" as input
+	for _, cmd := range lqlQueryCommands {
```

Comment:
Interesting approach, I am ok adding a common flag to multiple commands in this way, though we do follow this other way:

https://github.com/lacework/go-sdk/blob/main/cli/cmd/vuln_container.go#L259-L293

Totally up to you to update it or not. 😉

Reply:
Seems legit.  Let me give a a spin.

Reply:
Fixed.